### PR TITLE
Add DID static validation without tests

### PR DESCRIPTION
**Pull request by askolesov**
Signed-off-by: Andrew Nikitin <andrew.nikitin@cheqd.io>

### CODE REVIEW

**Comment by ankurdotb** [comment]
Task linked: [DEV-705 [performance] Split validation into basic and dynamic](https://app.clickup.com/t/6600954/DEV-705)

File: x/cheqd/types/did_validation.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+// That for groups:
+// Example: did:cheqd:testnet:fafdsffq11213343/path-to-s/ome-external-resource?query#key1???#123
+// 1 - [^/?#]* - all the symbols except / and ? and # . This is the DID part                      (did:cheqd:testnet:fafdsffq11213343)
+// 2 - [^?#]*  - all the symbols except ? and #. it means te section started from /, path-abempty (/path-to-s/ome-external-resource)
+// 3 - \?([^#]*) - group for `query` part but with ? symbol 									  (?query)
+// 4 - [^#]*     - group inside query string, match only exact query                              (query)
+// 5 - #([^#]+[\$]?) - group for fragment, starts with #, includes #                              (#key1???)
+// 6 - [^#]+[\$]?    - fragment only															  (key1???)
+// {0,1} - means that number of fragments can be only 0 or 1
+// Amount of query is not limited.
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:21]:
```suggestion
// Number of queries is not limited.
```

File: x/cheqd/types/did_validation.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	"github.com/cheqd/cheqd-node/x/cheqd/utils"
+	"regexp"
+)
+
+var SplitDIDRegexp, _        = regexp.Compile(`did:([^:]+?)(:([^:]+?))?:([^:]+)$`)
+var DidNamespaceRegexp, _    = regexp.Compile(`^[a-zA-Z0-9]$`)
+// Base58 only allowed (without OolI and 0)
+var UniqueIDRegexp, _        = regexp.Compile(`^[123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz]*$`)
+// That for groups:
+// Example: did:cheqd:testnet:fafdsffq11213343/path-to-s/ome-external-resource?query#key1???#123
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:13]: Is this DID valid? Can fragment be `#key1???#123`?

File: x/cheqd/types/did_validation.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+// 1 - [^/?#]* - all the symbols except / and ? and # . This is the DID part                      (did:cheqd:testnet:fafdsffq11213343)
+// 2 - [^?#]*  - all the symbols except ? and #. it means te section started from /, path-abempty (/path-to-s/ome-external-resource)
+// 3 - \?([^#]*) - group for `query` part but with ? symbol 									  (?query)
+// 4 - [^#]*     - group inside query string, match only exact query                              (query)
+// 5 - #([^#]+[\$]?) - group for fragment, starts with #, includes #                              (#key1???)
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:18]: In the example fragment is `#key1???#123`, here it's `#key1???`. Where is the truth?

File: x/cheqd/types/did_validation.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+	// Example: did:cheqd:testnet:base58str1ng1111
+	// match [0] - the whole string
+	// match [1] - cheqd                - method
+	// match [2] - :testnet
+	// match [3] - testnet              - namespace
+	// match [4] - base58str1ng1111     - id
+	match := SplitDIDRegexp.FindStringSubmatch(did)
+	if len(match) > 0 {
+		return match[1], match[3], match[4]
+	}
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:152]: Let's throw an error otherwise

File: x/cheqd/types/did_validation.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+func ValidateDID(did string, method string, allowedNamespaces []string) error {
+	method, namespace, unique_id := SplitDID(did)
+
+	// check method
+	if method != method {
+		return ErrStaticDIDBadMethod.Wrap(method)
+	}
+	// check namespaces
+	if !DidNamespaceRegexp.MatchString(namespace) || !utils.Contains(allowedNamespaces, namespace) {
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:114]: Let's move regexp matches to the split method

File: x/cheqd/types/did_validation.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+		return ErrStaticDIDNamespaceNotAllowed.Wrap(namespace)
+	}
+	// check unique-id
+	err := ValidateUniqueId(unique_id)
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:118]: This one too

File: x/cheqd/types/did_validation.go (hunk 5 above)
Comment (askolesov) [reply x/cheqd/types/did_validation.go:114]: So validate will only compare namespace / method with passed parameters

File: x/cheqd/types/did_validation.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+	return nil
+}
+
+func IsValidDID(did string, method string, allowedNamespaces []string) bool {
+	err := ValidateDID(did, method, allowedNamespaces)
+	return err == nil
+}
+
+// SplitDID panics if did is not valid
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:141]: Don't panic :) Return an error

File: x/cheqd/types/did_validation.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
+package types
+
+import (
+	"github.com/cheqd/cheqd-node/x/cheqd/utils"
+	"regexp"
+)
+
+var SplitDIDRegexp, _        = regexp.Compile(`did:([^:]+?)(:([^:]+?))?:([^:]+)$`)
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:8]:
```suggestion
var SplitDIDRegexp, _        = regexp.Compile(`^did:([^:]+?)(:([^:]+?))?:([^:]+)$`)
```

File: x/cheqd/types/did_validation.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+	if len(match) > 0 {
+		return match[1], match[3], match[4]
+	}
+
+	return "", "", ""
+}
+
+// SplitDIDUrl panics if did cannot be splitted properly
+func SplitDIDUrl(didUrl string) (did string, path string , query string, fragment string) {
+	match := SplitDIDURL.FindStringSubmatch(didUrl)
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:159]: Let's check for matches count

Comment (askolesov) [reply x/cheqd/types/did_validation.go:159]: Shouldn't we check for matches count?

File: x/cheqd/types/did_validation.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+func ValidateDIDUrl(didUrl string, method string, allowedNamespaces []string) error {
+	did, path, query, fragment := SplitDIDUrl(didUrl)
+	// Validate DID
+	err := ValidateDID(did, method, allowedNamespaces)
+	if err != nil {
+		return err
+	}
+	// Validate path
+	err = ValidatePath(path)
```
Comment (askolesov) [inline x/cheqd/types/did_validation.go:177]: And let's move those checks to the split method as well

**Review by askolesov — COMMENTED** [review]
This is a rocket, really. Just a few changes to apply

File: x/cheqd/types/did_validation.go (hunk 2 above)
Comment (anikitinDSR) [reply x/cheqd/types/did_validation.go:13]: no, it's not valid. I'll change it.

File: x/cheqd/types/did_validation.go (hunk 5 above)
Comment (anikitinDSR) [reply x/cheqd/types/did_validation.go:114]: Will it be cool to throw the exception while splitting the DID? From my point of view it's a phase of validation.

File: x/cheqd/types/did_validation.go (hunk 9 above)
Comment (anikitinDSR) [reply x/cheqd/types/did_validation.go:159]: It matches only once for Golang regexps... I checked it

Comment (anikitinDSR) [reply x/cheqd/types/did_validation.go:159]: BTW, The last changes in regexp doesn't match in case of fragment being `#key123???#####`

**Pull request merged**